tests/test_load_images.py

```diff
@@ -0,0 +1,143 @@
+        self.tmpdir.cleanup()
+
+    def testRelionSourceInOrder(self):
+        # ensure that we can load images in order using
+        # the start and num arguments
+        imgs = self.src.images[:500]
+        from_mrc = self.getParticlesFromIndices([i for i in range(0, 500)])
+        self.assertTrue(np.array_equal(imgs.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceNonsequential(self):
+        # ensure that we can load particles from different mrc files,
+        # nonsequentially, and end up with the ordering we asked for
+        indices = [501, 502, 503, 504, 505, 0, 1, 2, 3, 4, 729, 728, 730, 720]
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices(indices)
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceEveryOther(self):
+        even_indices = [i for i in range(0, self.n, 2)]
+        odd_indices = [i for i in range(1, self.n, 2)]
+        even_from_src = self.src.images[even_indices]
+        even_from_mrc = self.getParticlesFromIndices(even_indices)
+        odd_from_src = self.src.images[odd_indices]
+        odd_from_mrc = self.getParticlesFromIndices(odd_indices)
+        self.assertTrue(
+            np.array_equal(even_from_src.asnumpy(), even_from_mrc.asnumpy())
+        )
+        self.assertTrue(np.array_equal(odd_from_src.asnumpy(), odd_from_mrc.asnumpy()))
+
+    def testRelionSourceNDArray(self):
+        indices = np.arange(self.n)
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices(list(indices))
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceRange(self):
+        indices = range(self.n)
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices(list(indices))
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceFilter(self):
+        indices = filter(lambda x: x % 2 == 0, list(range(self.n)))
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices(
+            [i for i in range(self.n) if i % 2 == 0]
+        )
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceTuple(self):
+        indices = (1, 2, 3, 4, 5)
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices([1, 2, 3, 4, 5])
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testBadInput(self):
+        # test a non-iterable, non-NumPy, non-integer input
+        with self.assertRaisesRegex(KeyError, "Key for .images"):
+            _ = self.src.images[3.14]
+
+    def testBadNDArray(self):
+        with self.assertRaisesRegex(
+            KeyError, "Only one-dimensional indexing is allowed for images."
+        ):
+            _ = self.src.images[np.zeros((3, 3))]
+
```

Comment:
Also test for a `slice` input (with non-unit stride, indices outside of range, etc.)?

Reply:
What I did was add a type check for the `slice` components, and then a final check right before calling `_images` to find out of range indices that might have been in `filter`, `range` or `list` arguments.